### ISSUE TITLE
LauncherLayoutInfo should not be hardcoded

### ISSUE THREAD

**Issue by warpech** [Starcounter/starcounter-include#30]
Right now, `starcounter-include` reads the `LauncherLayoutInfo` from the merged JSON. That property is created by CompositionProvider ([here](https://github.com/Starcounter/CompositionProvider/blob/d7ae9f6d38cb4144a808a45ec905eca7dd9ef370/src/CompositionProvider/Program.cs#L23)) and it is assumed to be present in the merged JSON by the same name.

However, after https://github.com/Starcounter/level1/issues/4061 is released, the property appears in the merged JSON as `<CallerAppName>_<Index>__LauncherLayoutInfo_<Index>`, for example: `WebsiteProvider_0__LauncherLayoutInfo_0` ([full example](https://gist.github.com/warpech/2b82ffe56272ec8137d017543f115e79))

I am trying to search what's the good strategy to un-hardcode LauncherLayoutInfo. The only solution that comes to mind is to filter through all the partial namespaces until we find an object with the property `PartialId`.

@tomalec WDYT?

**Comment by tomalec** [comment]
> I am trying to search what's the good strategy to un-hardcode LauncherLayoutInfo. The only solution that comes to mind is to filter through all the partial namespaces until we find an object with the property PartialId.

Sounds like an overhead. Since we made the JSON ugly as hell already, maybe we could have just one namespace with metadata, with common name? I CallerAppName really needed there?

**Comment by warpech** [comment]
It is not a question for this thread but for https://github.com/Starcounter/level1/issues/4061, I will respond there.

**Comment by warpech** [comment]
@tomalec WDYT? I think that unhardcoding the namespace is still a good idea, because then it becomes forward compatible.

It would be intriguing to move the blending stuff from `starcounter-include` to `CompositionProvider`, but the only ways that I can think of are hacky. 

Like: 

1. every app can have a `Html` property that we use in https://github.com/Starcounter/starcounter-include/blob/30cf445ee88f4c44d57b2fef28a46729c498f780/starcounter-include.html#L253-L259

2. if any app (here I am thinking about `CompositionProvider`) has a `CompositionHtml` property, then it is used as the result of `buildUrl`, instead of building the URL on the client.

**Comment by tomalec** [comment]
I do agree it would be nice to un-hard-code it.

 I just think the way to discover it proposed in PR #31 is kind of hacky

> It would be intriguing to move the blending stuff from starcounter-include to CompositionProvider, but the only ways that I can think of are hacky.

I would rather not do so. We consider blending a Starcounter feature and we planned to incorporate CompositionProvider features in Starcounter eventually.

In past we already have changed our approach quite many times from "blending is always enabled" (`sc-include`) to "blending is only available if app dev wants it" (`imported-template` vs `launcher-include`) and we also had many problems with later, which led us to merge `launcher-include` to `starcounter-include`.


Correct me if I'm wrong, but we agreed on the "blending is always enabled, app dev have means to make his UI look pretty on first load, suite designer/composer/blender/editor (person) decides if, what and how to blend.

That's why personally I think it's not that bad to have fixed property names for starcounter-mapping-n-blending-related features.

Besides the fact it definitely should not be called "**Launcher**LayoutInfo".

**Comment by warpech** [comment]
You have valid points, but then the only alternative is to keep PR #31 but change the lines https://github.com/Starcounter/starcounter-include/blob/6f1edf1df7a78609badd2c848b419c31ca88b077/starcounter-include.html#L320-L326 to be:

```js
function findCompositionProvider(obj) {
    return obj.LauncherLayoutInfo || obj.CompositionProvider || obj.CompositionProvider_0;
};
``` 

WDYT?

**Comment by tomalec** [comment]
LGTM

**Comment by warpech** [comment]
Ok, then let's go with that. Could you please finish the PR #31 with that?

**Comment by tomalec** [comment]
ok

**Comment by tomalec** [comment]
`sc-include` does not reac (change composition and partial id) when `partial.CompositionProvider` object is changed, as it does for `LauncherLayoutInfo` https://github.com/Starcounter/starcounter-include/blob/master/starcounter-include.html#L210

**Comment by tomalec** [comment]
fixed by #41
